Trackerlist: Add toggle columns menu

thalieht opened the pull request:
And some coding style here and there.
And set alignment of columns with numbers to the right.

I wanted to fix coding style in whole trackerlist.cpp but the PR would get too big.
If sledgehammer doesn't have a problem with reviewing an extra commit i'll gladly do it.

thalieht commented:
Also what do you think about aligning the number columns to the right?
<img width="530" alt="tracker_align" src="https://cloud.githubusercontent.com/assets/6451685/23797398/5bec1cdc-05a8-11e7-9e46-d631cc8ae124.PNG">

Or it doesn't matter because these numbers have no units?

Chocobo1 commented on src/gui/properties/trackerlist.cpp:
```diff
@@ -57,25 +57,40 @@ TrackerList::TrackerList(PropertiesWidget *properties)
 {
+  // Set header
+  // Must be set before calling loadSettings() otherwise the header is reset on restart
+  setHeaderItem(new QTreeWidgetItem(headerLabels()));
```
what about `setHeaderLabels(const QStringList & labels)` ?

thalieht replied:
Sure but why is this comment and the other 2 below it flagged as outdated?

Chocobo1 commented on src/gui/properties/trackerlist.cpp:
```diff
@@ -529,13 +543,61 @@ void TrackerList::showTrackerListMenu(QPoint) {
+
+QStringList TrackerList::headerLabels()
+{
+    QStringList header;
```
style change please:
```c++
QStringList header {
    "#"
    ,tr("URL")
    ,...
};
```

Chocobo1 commented on src/gui/properties/trackerlist.cpp:
```diff
@@ -529,13 +543,61 @@ void TrackerList::showTrackerListMenu(QPoint) {
     Preferences::instance()->setPropTrackerListState(header()->saveState());
 }
+
+QStringList TrackerList::headerLabels()
```
`QStringList TrackerList::headerLabels() const`

thalieht replied:
You mean in the declaration?

Chocobo1 commented on src/gui/properties/trackerlist.cpp:
```diff
@@ -529,13 +543,61 @@ void TrackerList::showTrackerListMenu(QPoint) {
+
+    // Call menu
+    QAction *act = hideshowColumn.exec(QCursor::pos());
+    if (act) {
```
`if (!act) return;`

glassez replied:
Or just
```c++
if (!act) return;
```
Using pointer in boolean context is valid AFAIK (unlike int).

Chocobo1 commented on src/gui/properties/trackerlist.cpp:
```diff
@@ -529,13 +543,61 @@ void TrackerList::showTrackerListMenu(QPoint) {
+    int visibleCols = 0;
+    for (unsigned int i = 0; i < COL_COUNT; i++) {
+        if (!isColumnHidden(i))
+            visibleCols++;
```
just my opinion: prefer prefix increment in for-loops and here.

Chocobo1 commented on src/gui/properties/trackerlist.cpp:
```diff
@@ -529,13 +543,61 @@ void TrackerList::showTrackerListMenu(QPoint) {
+    }
+
+    // Call menu
+    QAction *act = hideshowColumn.exec(QCursor::pos());
```
`QAction *act = hideshowColumn.exec();`?

Chocobo1 commented on src/gui/properties/trackerlist.cpp:
```diff
@@ -529,13 +543,61 @@ void TrackerList::showTrackerListMenu(QPoint) {
+    QAction *act = hideshowColumn.exec(QCursor::pos());
+    if (act) {
+        int col = actions.indexOf(act);
+        Q_ASSERT(col >= 0);
```
this one is redundant

Chocobo1 commented on src/gui/properties/trackerlist.cpp:
```diff
@@ -529,13 +543,61 @@ void TrackerList::showTrackerListMenu(QPoint) {
+        actions.append(myAct);
+    }
+    int visibleCols = 0;
+    for (unsigned int i = 0; i < COL_COUNT; i++) {
```
suggestion: move this to a function
```c++
int TrackerList::visibleColumnsCount() const
```

evsh replied:
Why looping twice? You call `isColumnHidden()` in the previous loop already.

thalieht replied:
In a later commit it's moved in a seperate method in order to use it in the constructor as well. Take a look and let me know if you have a better idea.

evsh replied:
Pardon, did not see that. Well, I do have an idea, but doubt it worth implementation efforts. I would add a
```c++
template<class Func>
void for_each_column(Func f)
{
    for (int i = 0; ...) {
         f(i);
    }
}
```

thalieht replied:
I better not, i don't know anything about templates :)

Chocobo1 commented on src/gui/properties/trackerlist.cpp:
```diff
@@ -57,25 +57,40 @@ TrackerList::TrackerList(PropertiesWidget *properties)
   setSelectionMode(QAbstractItemView::ExtendedSelection);
+  header()->setStretchLastSection(false); // Must be set after loadSettings() in order to work
+  // Ensure that at least one column is visible at all times
+  bool atLeastOne = false;
```
then you can use `TrackerList::visibleColumnsCount()` here:
```c++
bool atLeastOne = visibleColumnsCount() > 0;
```

thalieht commented:
Thanks for the comments guys. Also noted for some future refactor because i pretty much copy pasted the same `displayToggleColumnsMenu()` in peerlist and searchtab.

Chocobo1 commented on src/gui/properties/trackerlist.cpp:
```diff
@@ -581,21 +575,36 @@ void TrackerList::saveSettings() const
+            ++visibleCols;
+
+        if (visibleCols > 1)
+            break;
```
this doesn't match the function name, you should remove it or give the function a more precise name.

thalieht replied:
I was thinking about this too but i can't think of an appropriate name so i'll remove it.

Chocobo1 commented on src/gui/properties/trackerlist.cpp:
```diff
@@ -69,19 +69,13 @@ TrackerList::TrackerList(PropertiesWidget *properties)
-  }
-  if (!atLeastOne)
+  bool atLeastOneVisible = visibleColumnsCount() > 0;
+  if (!atLeastOneVisible)
```
you can further reduce it:
`if(visibleColumnsCount() == 0)`

Chocobo1 commented on src/gui/properties/trackerlist.cpp:
```diff
@@ -340,15 +361,24 @@ void TrackerList::loadTrackers() {
+
+    item->setTextAlignment(COL_SEEDS, (Qt::AlignRight | Qt::AlignVCenter));
+    item->setTextAlignment(COL_PEERS, (Qt::AlignRight | Qt::AlignVCenter));
+    item->setTextAlignment(COL_DOWNLOADED, (Qt::AlignRight | Qt::AlignVCenter));
```
urgh... spamming everywhere... you should do it only in the constructor.

Chocobo1 replied:
why you need it twice? move it out of the `#if` block.

thalieht replied:
Ok i was thinking of something that isn't true :)

Chocobo1 commented on src/gui/properties/trackerlist.cpp:
```diff
@@ -296,6 +309,13 @@ void TrackerList::loadStickyItems(BitTorrent::TorrentHandle *const torrent) {
+    pex_item->setTextAlignment(COL_SEEDS, (Qt::AlignRight | Qt::AlignVCenter));
+    pex_item->setTextAlignment(COL_PEERS, (Qt::AlignRight | Qt::AlignVCenter));
+    lsd_item->setTextAlignment(COL_SEEDS, (Qt::AlignRight | Qt::AlignVCenter));
+    lsd_item->setTextAlignment(COL_PEERS, (Qt::AlignRight | Qt::AlignVCenter));
```
you should do it in the constructor

thalieht replied:
Ok

thalieht commented on src/gui/properties/trackerlist.cpp:
```diff
@@ -529,13 +543,61 @@ void TrackerList::showTrackerListMenu(QPoint) {
+{
+    QMenu hideshowColumn(this);
+    hideshowColumn.setTitle(tr("Column visibility"));
+    QList<QAction*> actions;
```
What?

evsh replied:
Instead of using the additional list with actions, you can put the index inside of the action object:
```C++
QAction *myAct = hideshowColumn.addAction(headerLabels().at(i));
myAct->setCheckable(true);
myAct->setChecked(!isColumnHidden(i));
myAct->setData(i);
```
and later instead of 
```C++
int col = actions.indexOf(act);
```
you would do:
```c++
int col = act->data().toInt();
```

thalieht replied:
Ok.

/Offtopic you probably missed my comment about static.

evsh replied:
Ops, sorry. Yes, I meant that that function can be static class member.

thalieht commented:
I saw in another PR that uncrustify did the opposite of this
```
-class PropertiesWidget: public QWidget, private Ui::PropertiesWidget
+class PropertiesWidget : public QWidget, private Ui::PropertiesWidget
```
Should i undo these?

thalieht commented on src/gui/properties/propertieswidget.h:
```diff
@@ -55,15 +55,18 @@ class QAction;
 QT_END_NAMESPACE
 
-class PropertiesWidget: public QWidget, private Ui::PropertiesWidget
+class PropertiesWidget : public QWidget, private Ui::PropertiesWidget
```
I saw in another PR that uncrustify did the opposite of this
Should i undo these?

evsh replied:
It seems that the Coding Guidlines do not say a word on this particular case. If so, it's up to you.

thalieht replied:
Although i prefer it with the extra space if you don't intend to change it in uncrustify i'll undo it for consistency.

evsh replied:
Our current uncrustify config instructs to remove the space (sp_before_class_colon), because I like it tight. We can discuss it, just open an issue.

evsh replied:
No, I was wrong, that particular parameter was set earlier by @sledgehammer999.

thalieht commented:
Added some coding style fixes to the corresponding commits and moved commit 
`Trackerlist: Allow to toggle columns` in the end so reviewers won't have to re-read the subsequent commits after fixups.

thalieht commented:
Rebased.

glassez commented:
@thalient, it is time for squashing fixup commits. Let's finish it if there are no more comments added.

thalieht commented:
Done.

glassez commented on src/gui/properties/trackerlist.h:
```diff
@@ -31,60 +31,78 @@
+    QList<QTreeWidgetItem *> getSelectedTrackerItems() const;
+
+private:
+    PropertiesWidget *properties;
```
One more requests. Can you fix coding style of these private names? Please change to camelCase and add `m_` prefix.

thalieht replied:
Sure

glassez commented on src/gui/properties/trackerlist.cpp:
```diff
@@ -524,13 +556,65 @@ void TrackerList::showTrackerListMenu(QPoint) {
+
+QStringList TrackerList::headerLabels()
+{
+    QStringList header {
```
Why not static?

Chocobo1 replied:
or remove this method and put the following to top of the file instead:
```c++
const QStringList headerLabels = {
   ...
}
```
and use `headerLabels[i]` below

@thalieht your choice

thalieht replied:
I don't know, i guess i should.

glassez replied:
@Chocobo1, then we can't use tr() (static QObject::tr produces translations out of context).

Chocobo1 replied:
>then we can't use tr() 

I see, it should be `QT_TRANSLATE_NOOP` in this way... it's a bit ugly.
(I'm not insisting on this)

glassez replied:
So `static` (or even `static const`) is best option here.

thalieht commented:
Added `static const` to `QStringList header` and fixed coding style of private members.

P.S. If you prefer the DHT in `m_DHTItem` etc. as lower case i'll change them.

glassez commented:
I prefer abbreviations in original case (as you did now).

glassez commented:
>I prefer abbreviations in original case (as you did now).

In cases where this notation is possible.

LordNyriox commented on src/gui/properties/trackerlist.h:
```diff
@@ -31,60 +31,78 @@
+        COL_PEERS,
+        COL_DOWNLOADED,
+        COL_MSG,
 
```
@thalieht:  I believe this blank line needs to be removed as well.

thalieht replied:
I don't know.. it separates the columns from their counter. Besides i've seen it elsewhere too. I can remove it if it's unwanted.

glassez replied:
Don't do it.
>it separates the columns from their counter.

Yes. It's a good idea to do so.

LordNyriox replied:
@thalieht:  I thought as much, but I was worried about style consistency.  

If that is not a problem, then you can leave it as is.

LordNyriox commented:
@thalieht:  I almost forgot… Do you think you could add in the features of the patch you referenced in [<https://github.com/qbittorrent/qBittorrent/pull/3162#issuecomment-272684219>]?  

Thanks.

LordNyriox commented:
@thalieht:  I almost forgot… Do you think you could add in the features of the patch you referenced in [<https://github.com/qbittorrent/qBittorrent/pull/3162#issuecomment-272684219>]?  

> ngosang here's a little patch that:
> 
> Separates seeds from peers for DHT PeX LSD
> Changes zeroed values that can't change to empty strings (if you see a zero you expect that it can change).
> Some coding style.
> If you want, take whatever you want from it :)
> 
> <https://gist.github.com/thalieht/14ae3106ae89f82d71a2a2c9f9ad29ed>

Thanks.

thalieht commented:
@LordNyriox the first is already in 3.3.11 #6385 
The second... i had hardcoded it (unpublished) but then i decided it'd be better to do it via `Hide zero and infinity values`. Maybe in some other PR... if i figure it out :)

LordNyriox commented:
@evsh, @glassez, @Chocobo1:  Perhaps this could be merged sometime soon?

LordNyriox commented:
@evsh, @glassez, @Chocobo1:  Perhaps this could be merged sometime soon?

evsh commented:
@Chocobo1, are you OK with these changes?

Chocobo1 commented:
>@Chocobo1, are you OK with these changes?

yeah, I have already approved it earlier (was dismissed by updates).

LordNyriox commented:
@evsh, @glassez, @Chocobo1:  I see formal approvals from all three of you.  

Is there any particular reason this PR has not yet been merged?

Chocobo1 commented:
Thank you!

thalieht commented:
Thanks
